### DISCUSSION TITLE
Permissions in a file

### DISCUSSION THREAD

**Discussion by lemyskaman** [orchidsoftware/platform#1523] [Ideas]
Hi guys, what you think about putting the orchid permissions in a file or a group of files. To be used by the orchid platform and orchid based apps in a similar fashion of laravel's config files.

So my idea is to allow the OrchidServiceProvider to loop all files within a directory (perhaps we might call it Permissions and put it inside app folder)  and in that loop read the multidimensional array inside each file an array that perhaps might look like this:

```php
        //App/Permissions/somepermissionfile.php
        
        return [
            'Systems Users'=>[   //the name of the permissions group
                'platform.systems.users.create' => 'Create',
                .
                .
                .
            ],
            .
            .
            .
        
        ];
```
As you can see in the above array scheme you can specify the group name that wraps the permission keys and inside of it you get the key pair permission/name.

The intention to allow orchid readd multiple files is to separate orchid needed permission that ships already for users roles main and system or examples, from what ever other permissions one might need related to the app

**Comment by osbre** [comment]
It's more familiar to store that kind of files in the `config` folder

**Reply by lemyskaman** [reply]
Familiar? well as you might have more than one permission file (only permission related files), I think its not a good idea to mix permission files with config files in the same folder

although you might be talking about only one permission file, in that case, it makes vanish the idea to keep separated the orchid related permissions, form orchid based apps permissions, and/or orchid based packages permissions.

**Reply by dhb52** [reply]
IMHO, we cannot use translation function inside of config. Permission groups need translation.

**Reply by bilogic** [reply]
@lemyskaman I think your suggestion is similar to `routes/web.php`, definitely useful to have it in a file.
Perhaps under `routes/platformPermissions.php`?
After all, permission to access/crud almost always involves routes, correct?

**Comment by dhb52** [comment]
Pls take translation into account, and do we need to cache it for quicker loading?

**Reply by lemyskaman** [reply]
there is no need to include translations in the permission files array 

off course translations is taken in consideration:

Imagine that we have all the permissions in multidimensional array called` $permission_array` then we in the boot method of a service provider we might have something like :

        
```php        
        foreach ($permission_array as $groupname=>$values){
             //creating the permission group with its translated name....
            $permission_group=$dashboard->registerPermissions(ItemPermission::group(__($groupname)));
            
             //adding the permissions to the group with its translation name....
            foreach($values as $permission_key=>$name){
                $permission_group->addPermission($permission_key,__($name));
            };
        };
 ```
        
as you can see there was no need of include the translation in the permission file, although the translating is being done at provider method level, at the time of loading the permission inside the dashboard class object

**Reply by lemyskaman** [reply]
Good question about cache.  IMHO is not needed the above proposed is a [PHP foerach loop, they seams to be fast](https://www.phpbench.com/) unless you are handling more than a million of permissions keys 

Honestly I have not a clue about  a chaching technique that would work for permissions.  

 they are not being cached anyway right now, they must be directly typed "Harcoded" on the orchid service provider boot method.

**Comment by lemyskaman** [comment]
I've finally came out with a method, which does exactly what I proposed. 

```php
    protected function registerPermissions(Dashboard $dashboard): self
    {

        collect(File::files(app_path('Permissions')))
            ->map(function ($file) {
                return require_once $file->getPathname();
            })
            ->collapse()
            ->map(function ($values, $group_name) use ($dashboard) {
                foreach ($values as $perm_key => $name) {
                    $dashboard->registerPermissions(
                        ItemPermission::group($group_name)
                            ->addPermission(
                                $perm_key,
                                __($name)
                            )
                    );
                }
            });

        return $this;
    }
```

**Reply by jonatemps** [reply]
how display my permission in sreen?

**Comment by jonatemps** [comment]
how display my permission in sreen?

**Reply by lemyskaman** [reply]
HI !      what do you mean with displaying permission on screen  ?
